File: src/test/java/com/n26/model/StatisticsModelTest.java
package com.n26.model;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.n26.controller.StatisticsController;

import com.n26.service.StatisticsComputation;
import com.n26.service.serviceImpl.TransactionCacheImpl;
import org.hamcrest.Matcher;
import org.hamcrest.Matchers;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.springframework.boot.test.json.JacksonTester;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.math.BigDecimal;

import static org.mockito.ArgumentMatchers.any;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;

public class StatisticsModelTest {

    private Statistics statistics;
    private MockMvc mockMvc;
    private JacksonTester<Statistics> jsonTransaction;

    @InjectMocks
    StatisticsController statisticsController;

    @Mock
    private TransactionCacheImpl transactionCache;

    @Mock
    private StatisticsComputation statisticsComputation;

    @Rule
    public MockitoRule rule = MockitoJUnit.rule();

    @Before
    public void setData(){
        JacksonTester.initFields(this, new ObjectMapper());
        statistics = new Statistics(BigDecimal.valueOf(15.43),
                BigDecimal.valueOf(15.43),
                BigDecimal.valueOf(15.43),
                BigDecimal.valueOf(15.43),
                1L);
        mockMvc = MockMvcBuilders.standaloneSetup(statisticsController).build();
    }

    @Test
    public void sumFormatTest() throws Exception{
        checkFieldFormatString("$.sum","15.43");
    }

    @Test
    public void avgFormatTest() throws Exception{
        checkFieldFormatString("$.avg","15.43");
    }

    @Test
    public void maxFormatTest() throws Exception{
        checkFieldFormatString("$.max","15.43");
    }

    @Test
    public void minFormatTest() throws Exception{
        checkFieldFormatString("$.min","15.43");
    }

    @Test
    public void countFormatTest() throws Exception{
        checkFieldFormatNumber("$.count",1);
    }

    private void checkFieldFormatString(String field, String value) throws Exception{
        checkFieldFormat(field,Matchers.is(value));
    }

    private void checkFieldFormatNumber(String field, Number value) throws Exception{
        checkFieldFormat(field,Matchers.is(value));
    }

    private void checkFieldFormat(String field, Matcher matcher) throws Exception {
        Mockito.when(statisticsComputation.computeStatistics(any())).thenReturn(statistics);
        mockMvc.perform(get("/statistics")
                .contentType(MediaType.APPLICATION_JSON)
                .content(jsonTransaction
                        .write(statistics)
                        .getJson()))
                .andDo(print())
                .andExpect(MockMvcResultMatchers.jsonPath(field, matcher));
    }
}

File: src/test/kotlin/com/model/StatisticsModelTest.kt
package com.model

import org.springframework.test.web.servlet.MockMvc
import org.springframework.boot.test.json.JacksonTester
import com.controller.StatisticsController
import com.service.serviceImpl.TransactionCacheImpl
import com.service.StatisticsComputation
import org.mockito.junit.MockitoJUnit
import org.junit.Before
import com.fasterxml.jackson.databind.ObjectMapper
import org.hamcrest.Matcher
import org.hamcrest.Matchers
import org.junit.Rule
import org.junit.Test
import org.mockito.*
import org.mockito.junit.MockitoRule
import java.math.BigDecimal
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import kotlin.Throws
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import java.lang.Exception

class StatisticsModelTest {
    private val statistics: Statistics = Statistics(
        BigDecimal.valueOf(15.43),
        BigDecimal.valueOf(15.43),
        BigDecimal.valueOf(15.43),
        BigDecimal.valueOf(15.43),
        1L
    )

    private lateinit var jsonTransaction: JacksonTester<Statistics?>

    @InjectMocks
    lateinit var statisticsController: StatisticsController

    private lateinit var mockMvc: MockMvc

    @Mock
    private lateinit var transactionCache: TransactionCacheImpl

    @Mock
    private lateinit var statisticsComputation: StatisticsComputation

    @Rule
    @JvmField
    var rule: MockitoRule = MockitoJUnit.rule()

    @Before
    fun setData() {
        JacksonTester.initFields(this, ObjectMapper())
        mockMvc = MockMvcBuilders.standaloneSetup(statisticsController).build()
    }

    @Test
    @Throws(Exception::class)
    fun sumFormatTest() {
        checkFieldFormatString("$.sum", "15.43")
    }

    @Test
    @Throws(Exception::class)
    fun avgFormatTest() {
        checkFieldFormatString("$.avg", "15.43")
    }

    @Test
    @Throws(Exception::class)
    fun maxFormatTest() {
        checkFieldFormatString("$.max", "15.43")
    }

    @Test
    @Throws(Exception::class)
    fun minFormatTest() {
        checkFieldFormatString("$.min", "15.43")
    }

    @Test
    @Throws(Exception::class)
    fun countFormatTest() {
        checkFieldFormatNumber("$.count", 1)
    }

    @Throws(Exception::class)
    private fun checkFieldFormatString(field: String, value: String) {
        checkFieldFormat(field, Matchers.`is`(value))
    }

    @Throws(Exception::class)
    private fun checkFieldFormatNumber(field: String, value: Number) {
        checkFieldFormat(field, Matchers.`is`(value))
    }

    @Throws(Exception::class)
    private fun checkFieldFormat(field: String, matcher: Matcher<Any>) {
        Mockito.`when`(statisticsComputation.computeStatistics(MockitoHelper.anyObject()))
            .thenReturn(statistics)
        mockMvc.perform(
            MockMvcRequestBuilders.get("/statistics")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    jsonTransaction
                        .write(statistics)
                        .json
                )
        )
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.jsonPath(field, matcher))
    }
    object MockitoHelper {
        fun <T> anyObject(): T {
            Mockito.any<T>()
            return uninitialized()
        }
        @Suppress("UNCHECKED_CAST")
        fun <T> uninitialized(): T =  null as T
    }
}